target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/common.c

```diff
@@ -320,22 +320,20 @@ static int __init rtl83xx_mdio_probe(struct rtl838x_switch_priv *priv)
 		if (of_property_read_u32(dn, "reg", &pn))
 			continue;
 
-		pr_info("%s found port %d\n", __func__, pn);
 		phy_node = of_parse_phandle(dn, "phy-handle", 0);
 		if (!phy_node) {
 			if (pn != priv->cpu_port)
 				dev_err(priv->dev, "Port node %d misses phy-handle\n", pn);
 			continue;
 		}
 
-		pr_info("%s port %d has phandle\n", __func__, pn);
 		if (of_property_read_u32(phy_node, "sds", &priv->ports[pn].sds_num))
 			priv->ports[pn].sds_num = -1;
 		else {
 			pr_info("%s sds port %d is %d\n", __func__, pn,
 				priv->ports[pn].sds_num);
 		}
-		pr_info("%s port %d has SDS\n", __func__, priv->ports[pn].sds_num);
+		pr_debug("%s port %d has SDS %d\n", __func__, pn, priv->ports[pn].sds_num);
```

Comment:
This change also happens to solve the watchdog-related issue on some HPE 1920 devices. (I didn't test with debug messages enabled during boot, though, so adding something like `cond_resched` additionally may still be worth considering.)

Maybe it would make sense to split out all the non-device specific commits into separate pull requests?

Reply:
@janh please do a (github) suggestion with the `cond_resched` and I'll put it in.

As for your fixes and non-device specific fixes, Maybe/probably. I'm still cleaning up the patches, re-organizing them. I don't want to mess with the patches content wise too much. But there's 3 goals for me personally. My primary goal, get my switch support merged :) But for that, secondly, I need the 9300 enablement patches. Not sure if we can merge patches without any devices.

Thirdly, the churn for getting stuff merged in openwrt is quite slow; so having 3 MR's (bugfixes/cleanups, 9300 support, device support) might mean foreverness, even though it's the right approach of course). Also, I lack any non-9300 hardware, so validating and testing the 'fixes' would be very hard.

Reply:
I do have an Zyxel XGS1250-12-ZZ0101F, maybe I'll open it and see about this, guess I can run initramfs

Reply:
initramfs would already be very helpful!

Reply:
> so adding something like cond_resched additionally may still be worth considering.

I've been thinking more about this, so the pr_debug is probably just the trigger here, it causes more cpu load then expected. (duh) but then what really IS the culprit? a race/timeout condition in this piece of code? I'm happy the change of the pr_info fixes it; i'm not happy that I don't yet understand why.

Reply:
Are pr_debug() lines printed? If console output is disabled during boot, that can lead to considerable time (CPU cycle) savings.

Reply:
afaik not by default; you need to enable debug prints specifically (might even be during compile time).

I'd be more then happy to remove 80% of all those prints; but, as you know this is not my code :( so having some debugging will probably still be usefull for the forseable time; and boot-time is probably not that important atm?

Reply:
> I've been thinking more about this, so the pr_debug is probably just the trigger here, it causes more cpu load then expected. (duh) but then what really IS the culprit? a race/timeout condition in this piece of code? I'm happy the change of the pr_info fixes it; i'm not happy that I don't yet understand why.

This is just because `printk` can be really slow, as it [directly writes to the console, if possible](https://www.kernel.org/doc/html/latest/core-api/printk-basics.html#c.printk).

On the HPE devices, the console runs at 38400 baud (this is also the rate used by the bootloader). Before this change, there are 3 calls to `pr_info` for every port, and if you calculate how long it takes just to write all of that to the console the result matches the actual execution time of this loop pretty well.

> afaik not by default; you need to enable debug prints specifically (might even be during compile time).

Yes, by default `pr_debug` is a no-op. The kernel needs to be built with `DEBUG` defined or with the [dynamic debug feature](https://www.kernel.org/doc/html/latest/admin-guide/dynamic-debug-howto.html) enabled.

I just did a few tests on my HPE 1920-48G with `CONFIG_KERNEL_DYNAMIC_DEBUG`, and `dyndbg="file drivers/net/dsa/rtl83xx/common.c +p"` added to cmdline to enable all debug prints from that file. I also tested with the second VPE disabled (`CONFIG_MIPS_MT_SMP` / `CONFIG_NR_CPUS` in the kernel config), as that gives more predictable results.

In these tests, I haven't seen the watchdog being triggered by this loop. The debug prints take about 0.7 seconds, so there is still a chance that it could happen if the watchdog hasn't been serviced shortly before.

The debug output from `rtl83xx_netdevice_event` together with that from `rtl83xx_fib_event` actually was enough to occasionally trigger the watchdog, though. And I wouldn't be surprised if enabling additional debug prints would show even more cases of this issue.

(I think if the [patch series moving the actual console writes to separate kernel threads](https://lore.kernel.org/linux-kernel/YouKQw72H7y9EJQK@alley/) actually lands in the kernel at some point, that would fix any issues related to slow `printk`. It was originally merged for Linux 5.19, but ended up being reverted as it didn't work properly with some console drivers.)

Reply:
You'd think that a) there'd be a debug print, that would do cond_reschedule after printing a few characters ;)

but even so, that loop must be 'quite full' without the debug prints? but then the question is; where to cond_reschedule? just before the pr_debug? Then it might be worth to add a local macro sched_pr_debug; that does cond_reschedule(); pr_debug instead.

Tricky business for sure regardless. Can we not set the console speed to 115200 on those devices as well? Yes it is annoying that U-Boot and linux have different speeds; but better then having a slow console?

Ideally, just replace all those custom u-boots with mainline u-boot and proper console speeds :) a man can dream ...

Reply:
> but even so, that loop must be 'quite full' without the debug prints

I'm not sure, the `for_each_node_by_name(dn, "port")` loop just reads some data from the device tree. It seems to take about 0.45 ms to run for 52 ports here.

> but then the question is; where to cond_reschedule? just before the pr_debug? Then it might be worth to add a local macro sched_pr_debug; that does cond_reschedule(); pr_debug instead.

In the end this isn't really an issue specific to the switch driver, but a general kernel issue. I tested with all debug output in the kernel enabled (using `dyndbg="file * +p"`), and that triggered the watchdog just 3 seconds after it was initialized…

> Can we not set the console speed to 115200 on those devices as well? Yes it is annoying that U-Boot and linux have different speeds; but better then having a slow console?

I don't think this is a good option, as it really is quite annoying in practice. And there wouldn't be a clear separation between bootloader and Linux, as early Linux console output uses the same rate as the bootloader anyway.

Reply:
in short; we should add that cond_reschedule, in combination with that pr_debug (and a comment as to why then).
```suggestion
		/* The pr_debug below adds a lot of overhead, so as long as this pr is here, we need a reschedule */
		cond_reschedule();
		pr_debug("%s port %d has SDS %d\n", __func__, pn, priv->ports[pn].sds_num);
```